### ISSUE TITLE
Tracking issue for `#![feature(maybe_uninit_extra,const_maybe_uninit_write)]`

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#63567]
This is a tracking issue for the RFC "Deprecate `uninitialized` in favor of a new `MaybeUninit` type" (rust-lang/rfcs#1892).

This issue specifically tracks the following unstable methods:

```rust
impl<T> MaybeUninit<T> {
    pub unsafe fn assume_init_read(&self) -> T { ... }
    pub unsafe fn assume_init_drop(&mut self) { ... }
}
```
and it tracks const-ness of
```rust
impl<T> MaybeUninit<T> {
    pub const fn write(&mut self, val: T) -> &mut T { ... }
}
```

**Comment by RalfJung** [comment]
Latest status here:

I am beginning to think that `read` should probably be renamed to `read_init` or `read_assume_init` or so, something that indicates that this may only be done after initialization is complete.

**Comment by SimonSapin** [comment]
@RalfJung Should the same logic apply to `get_ref` and `get_mut`? https://github.com/rust-lang/rust/issues/63568

**Comment by RalfJung** [comment]
That depends on whether references to invalid data are UB or not. If yes, then certainly. If not, then maybe.

Currently they are UB but that's just because it is the more conservative choice.

**Comment by ejmahler** [comment]
What stands in the way of stabilizing this?

**Comment by RalfJung** [comment]
Hm, good question. I'd list these:

* Figuring out if the return type of `write` makes sense and is useful as-is
* Figuring out if these methods are useful enough to justify adding them to the API

**Comment by RalfJung** [comment]
See https://github.com/rust-lang/rust/issues/80376 for a potential concern around the return type of `write`.

@mgeier you mentioned this to be a potential problem for `write_slice_cloned`; do you think the same issue also applies to `write`?

**Comment by mgeier** [comment]
I don't know whether this is an actual problem, but yes, it has the same issue: [playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a8550bfc7bee03e81826d77e65663e36)

BTW, the underlying `assume_init_mut()` has the same issue: [playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=af3a5021f60c9904e7b332b54c695845)

In the latter case it is arguably worse, because the method name contains `assume_init`, which suggests to me that this "activates" the destructor, which it doesn't. But I guess that has a different tracking issue: #63568.

For me, the real question is: what are the semantics of assigning to a `&mut T`?

I assume that when I'm assigning a new value, the old one will be guaranteed to be dropped. But what about the new value? Is the new value guaranteed to be dropped at some point? If yes, this issue is an actual problem.

But AFAICT, it is allowed in safe Rust to call `std::mem::forget()` on anything, so a `&mut T` is simply unable to guarantee that an assigned value will ever be dropped.

Still, I'm not sure whether accidental leaking should be made that easy?

**Comment by SimonSapin** [comment]
We already have stable `impl DerefMut for ManuallyDrop` which makes this accidental leaking just as easy.

**Comment by RalfJung** [comment]
> In the latter case it is arguably worse, because the method name contains assume_init, which suggests to me that this "activates" the destructor, which it doesn't. But I guess that has a different tracking issue: #63568.

Why do you think `assume_init` has anything to do with destructors? That was never the intention. The only thing it is intended to convey is "if this is not yet initialized, there's UB".

**Comment by mgeier** [comment]
> Why do you think `assume_init` has anything to do with destructors? That was never the intention. The only thing it is intended to convey is "if this is not yet initialized, there's UB".

I guess the problem is that the function name `assume_init()` doesn't contain what it actually *does*.

Therefore, in my (limited) experience with stable Rust, I've come to imagine that `assume_init()` is in reality a shortcut that in its full form means "activate the destructor, assuming the memory has been properly initialized before".

That is fine as long as there is only a single function containing the words "assume init".
Now that multiple functions with "assume init" are proposed to be added, this becomes inconsistent and confusing.

I've looked back to the discussions about naming `assume_init()` (which happened before I was a Rust user) and I found https://github.com/rust-lang/rust/issues/53491#issuecomment-438595288 which argues "Putting both the emphasis that it has to be initialized and a description of what it does (unwrap/into_inner/etc.) into one name gets rather unwieldy".
I think this was a fine argument as long as there was only one function containing "assume init", but it doesn't work with multiple "assume init" functions.

Now that we have a stable function `assume_init()`, I think we have to accept that the name `assume_init` contains more meaning than simply the two words "assume init", and I think we should use the words "assume init" in newly added functions only if they carry the *full* meaning of the already existing `assume_init()` function.

**Comment by RalfJung** [comment]
I am rather surprised by this. Whether or not the destructor runs is fully encoded in the type -- when a `T` goes out of scope, its destructor runs; when a `MaybeUninit<T>` goes out of scope, nothing happens. This is the first time that I hear of the idea of "activating a destructor".

> Now that we have a stable function assume_init(), I think we have to accept that the name assume_init contains more meaning than simply the two words "assume init", and I think we should use the words "assume init" in newly added functions only if they carry the full meaning of the already existing assume_init() function.

I disagree -- I think having more of these methods will actually make this more clear. Once `assume_init_read` and `assume_init_ref/mut` are stable, it seems unlikely that people would associate `assume_init` with "activating drop".

**Comment by Coder-256** [comment]
I think in other words, the confusion is that `assume_init/`assume_init_read` move the value, meaning it will be dropped, but `assume_init_ref`/`assume_init_ref` only reference the value, meaning it will not be dropped. I see how this can be confusing but honestly I actually like design because it's pretty similar to how pointers already work: the former two are like `std::ptr::read()`, and the latter two are like casting from a pointer to a reference.

**Comment by mgeier** [comment]
@RalfJung 

> This is the first time that I hear of the idea of "activating a destructor".

Sorry that I'm using such naive and hand-wavey terminology.

But that's how I, as a humble Rust user, think about it. I have a `MaybeUninit<T>` variable which I, let's say initialize with some FFI function. Now the bits in memory have already the correct values, but in order to further work with it, I'd like to turn it into a `T`, which logically only does one thing: it "activates" the destructor.

> I think having more of these methods will actually make this more clear. Once `assume_init_read` and `assume_init_ref/mut` are stable, it seems unlikely that people would associate `assume_init` with "activating drop".

That's true, but also because `assume_init()` will be stripped of any meaning with regards to what it actually does.

But anyway, I'm wondering: why do we even need `assume_init_read` and `assume_init_ref/mut`?

Aren't they just wrappers for one-liners?

I have the feeling that adding all those function will make Rust more complicated without really gaining anything.

I think using the one-liners directly makes it much more obvious to see what actually happens in the code.

@Coder-256 

> I see how this can be confusing but honestly I actually like design because it's pretty similar to how pointers already work: the former two are like`std::ptr::read()`, and the latter two are like casting from a pointer to a reference.

I don't think this is similar at all. The function name `std::ptr::read()` contains a verb that tells me what the function actually does. In contrast, `assume_init()` does not tell me what it actually does (but granted, I can get quite a good idea by looking at the involved types).

Casting from a pointer to a reference isn't similar either. It is done by using sigils alone and doesn't involve any function names at all.

**Comment by RalfJung** [comment]
> That's true, but also because assume_init() will be stripped of any meaning with regards to what it actually does.

Well, the main meaning (I'd say) is that initialization is done and the compiler may now assume that the data is initialized. So I don't think meaning is "stripped".

But anyway, this is useful feedback, even if I don't quite know what to do with it. ;) Thanks!

**Comment by faern** [comment]
> ... which logically only does one thing: it "activates" the destructor.

That's not the only thing calling `assume_init` does. It gives you a `T`! You did not have a `T` until that point in the code. Meaning you can now suddenly have access to the fields and methods on `T`, something you did not have on the `MaybeUninit<T>`. At least not safely and easily.

**Comment by mgeier** [comment]
@RalfJung 

> > That's true, but also because assume_init() will be stripped of any meaning with regards to what it actually does.
> 
> Well, the main meaning (I'd say) is that initialization is done and the compiler may now assume that the data is initialized. So I don't think meaning is "stripped".

It is true that the name is supposed to signal that initialization has been done (before calling the function) and the compiler may assume that the data is initialized (after calling the function). But there is still no information about any "action" that the function might be doing during its runtime (which is fine as long as there is only one function containing `assume_init`).

@faern 

> you can now suddenly have access to the fields and methods on `T`

Thanks, I completely forgot about that!

But this is only tangential to the current issue anyway.

---

To come back to the original issue:

My opinion: one function named `assume_init()` is fine, adding more `assume_init_*()` functions is inconsistent.

My suggestions: don't add those new `assume_init_*()` functions.

**Comment by RalfJung** [comment]
> It is true that the name is supposed to signal that initialization has been done (before calling the function) and the compiler may assume that the data is initialized (after calling the function). But there is still no information about any "action" that the function might be doing during its runtime (which is fine as long as there is only one function containing assume_init).

The "action" is the same for all of these functions -- there is no action. They are all equivalent to `transmute`.

The fact that after `assume_init`, things get dropped, is a consequence of the general rules of Rust, not of any "action" that is being taken here. If we follow your argument, a `transmute` from `ManuallyDrop<T>` to `T` should also be considered to take an "action" such as "activating the destructor" -- I do not think that this is a good way to think about this. IMO the way to think about this is that the `assume_init` function does nothing, and that destructors get run when an initialized variable type with drop glue leaves scope. This is a much more compact way to represent all the same behavior, and it does not require learning for each function its "effect on dropping" -- there is none! Dropping is entirely orthogonal to all these functions.

**Comment by mgeier** [comment]
@RalfJung The `transmute()` function is actually a very good counter-example. Its name mentions an "action", namely the verb "to transmute". Therefore, it is no problem to have additional functions containing the word `transmute`, e.g. `transmute_copy()`.

**Comment by est31** [comment]
@mgeier 

> AFAICT, it is allowed in safe Rust to call std::mem::forget() on anything, so a &mut T is simply unable to guarantee that an assigned value will ever be dropped.

> Still, I'm not sure whether accidental leaking should be made that easy?

The only places that return mutable references are ones which have previously ensured that something has been written to them. From a conceptual point of view, if you overwrite the `&mut T` or write into the slice, you don't increase the number of leaked things, because at assignment, the drop function is called for the item that is being replaced. What are the alternatives if no mutable reference is available? If people use `MaybeUninit::write` or `ptr::write`, no destructor is called for the old item. So one could argue that *not* using `&mut T` can lead to leaks as well, maybe even more so if there are multiple opportunities of writing to the `MaybeUninit` value.

-------

How do people feel about a proposal to stabilize the `_extra` feature? I find the `write` function especially useful.

The example code in the [array_assume_init](https://doc.rust-lang.org/1.53.1/std/mem/union.MaybeUninit.html#method.array_assume_init) docs looks way nicer if you can use `write`:

```rust
let mut array: [MaybeUninit<i32>; 3] = MaybeUninit::uninit_array();
array[0] = MaybeUninit::new(0);
array[1] = MaybeUninit::new(1);
array[2] = MaybeUninit::new(2);
```

becomes

```rust
let mut array: [MaybeUninit<i32>; 3] = MaybeUninit::uninit_array();
array[0].write(0);
array[1].write(1);
array[2].write(2);
```

Overall, I feel rustdoc could be improved a little bit, but this can be done by a stabilization PR.

* In the safety section of `assume_init_drop`, I see an attempt to explain what I believe is the requirement that the data must not just be *initialized* but also *valid*. It's still a bit unclear what a "1-initialized Vec<T>" is. I'm sure it isn't `vec![1]`, but that's the initial thought when reading it.

* The example in `assume_init_read` explains very clearly how calling the function twice causes issues. The text leading up to the example is a bit terse on this question and should point out this footgun more explicitly. I feel the [`ptr::read`](https://doc.rust-lang.org/1.53.1/std/ptr/fn.read.html#ownership-of-the-returned-value) docs do this really well, and maybe one can copy over some text from them, maybe also a link.

* Lastly, `write` could use an example, possibly highlighting the double-write footgun.

**Comment by mgeier** [comment]
Is it possible to split this issue into two parts?

I think it would be good to be able to stabilize `MaybeUninit::write()` without being forced to stabilize `MaybeUninit::assume_init_read()` and `MaybeUninit::assume_init_drop()` as well.

The first function is a real help for avoiding `unsafe` code, but the latter ones are just as `unsafe` as the one-liners in their implementation, they pollute the documentation, and make Rust just a wee bit harder to learn and less joyous to use.

@est31 Your example mentions how `write()` makes for better code, but do you think that's true for the other two proposed functions as well?

> The only places that return mutable references are ones which have previously ensured that something has been written to them. From a conceptual point of view, if you overwrite the `&mut T` or write into the slice, you don't increase the number of leaked things, because at assignment, the drop function is called for the item that is being replaced.

It is true that "you don't increase the number of leaked things". My point is when I get a `&mut T` from *somewhere*, I might not know whether it points into a `MaybeUninit` or not. So I indeed don't increase the number of leaked things, but I also *don't know* the number of leaked things. And I might care about whether the `T` I'm assigning will be dropped eventually or not.

But to be clear (which I probably wasn't before): I still think that stabilizing `MaybeUninit::write()` in the currently proposed form is a good thing. Rust doesn't have a formal marker for "may leak", so I think this should be added to the docs, and everything is fine.

**Comment by est31** [comment]
@mgeier 

For your point on splitting the issue, I think it's a good idea to maybe only stabilize `write` first. I mainly care about the `write` function, and for the other two functions you do have a point. They are both one-liners. Yes, `write` is a one liner as well but it feels to me that its use case is way more common than the other two. To be honest, this `_read` function seems like a big footgun to me given that a function that *reads* causes destructors to run, but it shows up on grep.app, so there seem to be use cases for it.

> Rust doesn't have a formal marker for "may leak", so I think this should be added to the docs, and everything is fine.

Alright I'll incorporate it into #86220 .

**Comment by RalfJung** [comment]
> The first function is a real help for avoiding unsafe code

(the first function being `write`)
The example shown above certainly becomes a bit nicer with `write`, but it does not avoid any `unsafe`. So this advantage of `write` that you mention seems purely theoretic to me. Or are their real practical cases where `write` avoids `unsafe`?

**Comment by mgeier** [comment]
> So this advantage of `write` that you mention seems purely theoretic to me. Or are their real practical cases where `write` avoids `unsafe`?

Yes it is indeed purely theoretic, I personally have never been in a situation where I needed something like that.
I don't know of a practical situation where this actually matters.

I don't see myself ever needing the `write()` function, but if it is decided to stabilize it, would suggest to add a warning to the docs, which @est31 just did in ef6c9e653d605c83eca23bd4cd63ebfa8b17923e.

It may or may not be worth stabilizing the `write()` function, but I think the decision should be independent of the potential stabilization of `assume_init_read()` and `assume_init_drop()`.

**Comment by RalfJung** [comment]
> the decision should be independent of the potential stabilization of assume_init_read() and assume_init_drop().

That's fine, I lumped them into one feature gate solely because it wasn't at all clear how things would be staged best.

**Comment by est31** [comment]
For me, the killer feature of `write` is not that it's safe but that it's more convenient to use correctly. E.g. it's unsound to obtain a pointer to a MaybeUninit, turn that pointer into a reference, and then write to it. You are supposed to use the pointer apis to never have a reference lying around, but this is less convenient than the unsound solution. I think it's way easier if people can just use `write` instead, which is so non-dangerous to use, it's not even `unsafe`.

**Comment by est31** [comment]
I've made a proposal to stabilize the `write` function in isolation for now, and let the question on the other two functions open, see: #86344

**Comment by dtolnay** [comment]
Proposing FCP for just the `write` method, as suggested in https://github.com/rust-lang/rust/pull/86344.

```rust
impl<T> MaybeUninit<T> {
    pub fn write(&mut self, value: T) -> &mut T
}
```

This method is a **safe** equivalent to the **unsafe** `.as_mut_ptr().write(value)` + `assume_init_mut()`.

There is some discussion [above](https://github.com/rust-lang/rust/issues/63567#issuecomment-751401407) about whether the return type makes it unnecessarily prone to leak things (as opposed to returning `()` I guess), but I don't have a problem with this signature.

@rust-lang/libs-api
@rfcbot fcp merge

**Comment by joshtriplett** [comment]
This feels really late to be suggesting, and I apologize for the bikeshedding, but `write` feels like the wrong name for this method. `write` makes me think of the method of the same name in `io::Write` and `fmt::Write`, which feels confusing.

This method seems more like `set`, or perhaps `init`.

@rfcbot concern name-conflict-with-write-traits

**Comment by est31** [comment]
`set` would be good as well I think. It has prior art in the standard library in the form of `Cell::set`. `init` has no prior art that I know of. The other alternative would be `insert`, which [has prior art too](https://github.com/rust-lang/rust/pull/86344#issuecomment-862441264). Overall I think I still prefer `write` due to the similarity to `ptr::write`. Regarding the point about confusability, I think `MaybeUninit` belongs to the "unsafe" part of the std abstractions similar as `ptr::write`, while `io::Write` and `fmt::Write` belong to a higher level of "interacting with the outside".

**Comment by joshtriplett** [comment]
I can certainly acknowledge the similarity to `ptr::write`, but at the same time, I think the potential for confusion there seems high.

I'd be happy with `insert` as well, and I don't have any preference among `set` and `insert`. I only suggested `init` because the type itself is `MaybeUninit` and it has methods like `assume_init`, so an `init` method seems to fit.

What do other @rust-lang/libs-api folks think? If everyone else on the team feels that `write` is strictly preferable to any of `set`, `insert`, or `init`, I'll withdraw the objection and go with that consensus. But I'd love to know if there's a reason to *prefer* `write` over any of those; I personally think there are reasons to prefer any of those over `write`.

**Comment by RalfJung** [comment]
I agree having `write` in `ptr` as well as `io`/`fmt` is potentially confusing, but IMO that ship has long sailed. For `MaybeUninit` I think we'd need a good reason to *not* follow the precedent of `ptr` (like how we have `assume_init_read` instead of `read`).

**Comment by BurntSushi** [comment]
I'm with @RalfJung. The ship has sailed. And in particular, this is a convenience routine that replaces a use of pt::write, so I feel it's probably the right choice here.

**Comment by BurntSushi** [comment]
OK, so reading over the discussion above, there seems to be an implication that this routine is only theoretically useful for avoiding `unsafe`. Do we have some examples where this actually avoids `unsafe`?

Maybe this is a silly question, but why doesn't `write` return `T` instead of `&mut T`? If it did that, then you wouldn't need to later call `assume_init` to get the thing you wrote. (Perhaps a `write` that returns a `T` is an altogether different method though.)

**Comment by RalfJung** [comment]
> Maybe this is a silly question, but why doesn't write return T instead of &mut T?

`write` takes a `T`, how could it possibly be useful to return it again? (Not to mention that without `T: Copy`, this would be extremely foot-gunny.)

**Comment by BurntSushi** [comment]
@RalfJung Oops. I meant to also say, "and also take `self` instead of `&mut self`."

**Comment by dtolnay** [comment]
Taking `self` + `T` and returning `T` would be equivalent to the caller just dropping their MaybeUninit and doing whatever they want with the `T` that they already have, which is much easier without any method involved.

**Comment by BurntSushi** [comment]
@dtolnay Hmm, I see. So I guess what I'm left with is a question about when one might want to use this. @est31 gave an example above, but it didn't eliminate any `unsafe`. That's not to say that that means we shouldn't add this, but, for example, I would like to see a doc example for this routine showing when it's most useful.

**Comment by dtolnay** [comment]
Some use cases I found in GitHub search:

- https://github.com/strake/hash-table.rs/blob/25d89bc43a4f4a8a366eaccf67b5ce8041d29023/src/lib.rs#L106-L108 &mdash; this is dealing with `self.elems` of type `Es: IndexMut<usize, Output = MaybeUninit<(K, T)>>`

- https://github.com/maboesanman/cozal/blob/d25f626384f49451582b50d8a6fb5df44714cb4d/src/source/adapters/transposer/engine/pin_stack.rs#L72-L73 &mdash; pushing onto a `Vec<Box<[MaybeUninit<T>]>>` from a `&mut self` method and returning `Pin<&mut T>`

- https://github.com/quartiq/thermostat/blob/1fc3131baaf6dfd6b3e47f4be5e1d7187858d126/src/usb.rs#L27-L29 &mdash; initialization and use of a global MaybeUninit

I think these would all eliminate an `unsafe` by using `write`.

**Comment by BurntSushi** [comment]
All righty, thanks! This LGTM once we have a good doc example. @est31 Would it be possible to include one (perhaps based on one of @dtolnay's examples) in your stabilization PR?

@rfcbot concern doc-example

**Comment by joshtriplett** [comment]
@RalfJung wrote:
> For `MaybeUninit` I think we'd need a good reason to _not_ follow the precedent of `ptr` (like how we have `assume_init_read` instead of `read`).

OK, that's a fair point, especially given that we already have methods with `read` in the name.

Last question then: would it potentially make sense, given the precedent of `assume_init_read` rather than `read`, to similarly call this `init_write` or something along those lines?

**Comment by RalfJung** [comment]
> especially given that we already have methods with read in the name.

Note that `assume_init_read` is still unstable.

> Last question then: would it potentially make sense, given the precedent of assume_init_read rather than read, to similarly call this init_write or something along those lines?

Maybe? `assume_init` is how we mark all the methods that, well, assume that the data is initialized. This is a crucial part of their safety contract. `write` on the other hand is safe so this is less pressing.

But I could imagine this making sense -- I guess let's see what people that actually use this API (unlike me ;) say.

**Comment by SimonSapin** [comment]
`init_write` sounds like “initialize and write” to me, with two redundant almost-synonyms verbs. (Whereas `assume_init_read` is “read while assuming already initializ*ed*” with `init` describing a state rather than an action.)

Calling it `write` like the pointer method makes sense to me.

**Comment by joshtriplett** [comment]
Alright (all `write` ;) ), sounds like the consensus is to stick with `write` then. I withdraw the concern.

@rfcbot resolved name-conflict-with-write-traits

**Comment by est31** [comment]
> Would it be possible to include one (perhaps based on one of @dtolnay's examples) in your stabilization PR?

Will try to build a nice self contained example.

(As a note, the `hash-table.rs` example linked in the comment above is actually unsound as it calls `assume_init_mut` on an uninitialized value, yielding a mutable reference to it. I think this is further evidence that safe APIs for MaybeUninit are highly useful.)

**Comment by est31** [comment]
@BurntSushi is the example added by #86344 good? It bases on the pin_stack code example.

**Comment by BurntSushi** [comment]
@est31 Yes, love it, thank you. :-)

@rfcbot resolve doc-example

**Comment by Kixunil** [comment]
I'd like to report that I myself found `assume_init_*` helper methods very useful. `MaybeUninit` is likely to be used around other `unsafe` code, in my case I had *roughly* `NonNull<UnsafeCell<MaybeUninit<T>>>`. I hope I don't have to explain what kind of pointer-manipulation soup I had to write just to read `T` from that thing (after ensuring it's initialized, with proper synchronization of course).

I also found their names perfectly clear and natural. No dropping confusion as discussed above. Not only that, I intuitively expected those methods to exists and searched them in the docs. `assume_init_drop` was a happy surprise because it felt like a corner-case yet I found it. Ironically, I was only confused by (stable) `assume_init` as I mistook it for `assume_init_read`. Maybe `assume_init_into_inner` would have been more clear but that is not an option now.

I can definitely say I look forward to them being stabilized.

**Comment by Amanieu** [comment]
I'd like to move forward and stabilize this API. This covers both 

```rust
impl<T> MaybeUninit<T> {
    pub unsafe fn assume_init_read(&self) -> T { ... }
    pub unsafe fn assume_init_drop(&mut self) { ... }
}
```
and the const-ness of
```rust
impl<T> MaybeUninit<T> {
    pub const fn write(&mut self, val: T) -> &mut T { ... }
}
```

@rfcbot merge

**Comment by est31** [comment]
@Amanieu I think the command is "rfcbot fcp merge"

**Comment by Amanieu** [comment]
@rfcbot fcp merge

**Comment by Amanieu** [comment]
@rfcbot merge

**Comment by Amanieu** [comment]
@rfcbot fcp cancel

**Comment by Amanieu** [comment]
@rfcbot merge

**Comment by ojeda** [comment]
AFAICT, no PR was sent for this. Sending one for the `maybe_uninit_extra` bit.

**Comment by RalfJung** [comment]
Yes, there are other `assume_init` functions on `MaybeUninit`. They are all just memcpy's, but they are *type-changing* memcpys and that makes all the difference.

**Comment by Kixunil** [comment]
To me it's a natural combination of `ptr::read()` and `assume_init`.

**Comment by RalfJung** [comment]
That's blocked on stabilizing any use of `&mut` in `const`.

**Comment by DaniPopes** [comment]
Opened #116233 to const-stabilize `assume_init_read` since it looks like it was only blocked by `ptr::read` (https://github.com/rust-lang/rust/pull/92768#issuecomment-1011101383), which has since been stabilized in 1.71.
